docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1060 @@
+
+- Pageservers: a node generation ID is acquired at startup and lasts until the process ends or another
+  process with the same node ID starts.
+- Tenant attachments: for a given tenant, each time its attachment is changed, a per-tenant generation
+  number increases.
+
+This provides some important invariants:
+
+- If there are multiple pageservers running with the same node ID, they are guaranteed to have
+  a different generation ID.
+- If there are multiple pageservers running with the same node ID, we may unambiguously know which
+  of them "wins" by picking the higher generation number.
+- If two pageservers have the same tenant attached, they are guaranteed to have different attachment
+  generation numbers.
+
+The node generation number defines a global "in" definition for pageserver nodes: a node whose
+node generation matches the node generation the control plane most recently issued is a member of the cluster
+in good standing. Any node with any older generation ID is not.
+
+Distinction between generation numbers and Node ID:
+
+- The purpose of a node generation ID is to provide a stronger guarantee of uniqueness than a Node ID.
+- Node generation ID is guaranteed to be globally unique across space and time. Node ID is not unique
+  in the event of starting a replacement node after an original node is network partitioned, or
+  in the event of a human error starting a replacement node that re-uses a node ID.
+- Node ID is written to disk in a pageserver's configuration, whereas node generation number is
+  received at runtime just after startup (this does incur an availability dependency, see [availability](#availability)).
+- The two concepts could be collapsed into one if we used ephemeral node IDs that were only ever used one time,
+  but this makes it harder to efficiently handle the case of the same physical node restarting, where we may want
+  to retain the same attachments as it had before restart, without having to map them to a different node ID.
+- We may in future run the pageserver in a different cluster runtime environment (e.g. k8s) where
+  guaranteeing that the cluster manager (e.g. k8s) doesn't run two concurrent instances with the same
+  node ID is much harder.
+
+#### Why use per-tenant _and_ per-node generation numbers?
+
+The most important generation number is the tenant attachment number: this alone would be sufficient
+to implement safe migration and failover, if we assume that our control plane will never concurrently
+run two nodes with the same node ID.
+
+Running two nodes with the same ID might sound far-fetched, but it would happen very easily if we
+ran the pageserver in a k8s environment with a StatefulSet, as k8s provides no guarantee that
+an old pod is dead before starting a new one.
+
+The node generation is useful other ways, beyond it's core correctness purpose:
+
+- Including both generations (and the node ID) in object keys also provides some flexibility in
```

Comment:
> I assume this will go away soon?

This is the 16 bit node_id part of the suffix -- I think it will go away eventually, but it's cheap, and in the near term acts as an extra layer of protection if we had some bug with attachment logic (two nodes will never overwrite one another keys)